File: 98_validate_binary_search_tree/isValidBST.ts
function isValidBST(root: TreeNode | null): boolean {
  return dfs(root, -Infinity, Infinity);
}

function dfs(root: TreeNode | null, minVal: number, maxVal: number) {
  if (root == null) return true;

  if (root.val <= minVal || root.val >= maxVal) return false;

  return dfs(root.left, minVal, root.val) && dfs(root.right, root.val, maxVal);
}

File: 98_validate_binary_search_tree/isValidBST.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	return dfs(root, math.MinInt, math.MaxInt)
}

func dfs(root *TreeNode, minVal int, maxVal int) bool {
	if root == nil {
		return true
	}

	if root.Val <= minVal || root.Val >= maxVal {
		return false
	}

	return dfs(root.Left, minVal, root.Val) && dfs(root.Right, root.Val, maxVal)
}